### ISSUE TITLE
It'd be nice if the connection pool was grown in parallel and/or grown on demand

### ISSUE THREAD

**Issue by mikehearn** [robtimus/sftp-fs#10]
At the moment connecting to a server is quite slow by default because each connection in the pool is added serially, and it defaults to five connections. So that's like logging in 5 times before you can do anything.

Is it possible to make the pool expand on demand so the need to manually configure the pool size and wait for it to be built can be fixed?

**Comment by robtimus** [comment]
Expanding on demand won't be that hard. Shrinking it back to size on the other hand is a different issue. I don't want to let the pool size increase every time a connection is needed but none is available.

I was thinking about filling the pool with (mostly) purposefully broken connections. That could probably work, but the problem with that would be that if there's one working connection, that wouldn't be put back in the pool in such a way that it would be retrieved the next time. As a result, the first few requests would each lead to a new login, even if they're all done in the same thread.

I have an idea how to fix this, but I need to see how this works out. An alternative is to use something like Apache Commons Pool, but I always found that overkill. If I can't get my other idea to work that may still be the final solution.

**Comment by robtimus** [comment]
FYI, I've been working on this. Because the pool with actual Jsch connection was hard to test, I extracted that part to a separate project: https://github.com/robtimus/simple-pool. I've added 4 configuration settings, to keep it simple:

* The maximum size (the current client connection count)
* The initial size (new setting; default 1); this is what you actually wanted
* The maximum wait time (the current client connection wait timeout; 0 means don't wait now)
* The maximum idle time (new setting; default no maximum)

I am testing the two projects some more, and I also want to integrate the new project in https://github.com/robtimus/ftp-fs. Once that's done I will release a version 3.0. That will mostly be compatible with the current version, but I replaced the two previous settings with one setting for the pool configuration.
